lib/foreman_remote_execution_core/async_scripts/retrieve.sh

```diff
@@ -8,31 +8,78 @@ fi
 
 BASE_DIR="$(dirname $(readlink -f "$0"))"
 
-function release_lock() {
-    rm $BASE_DIR/retrieve_lock
-}
+# load the data required for generating the callback
+. "$BASE_DIR/env.sh"
+URL_PREFIX="$CALLBACK_HOST/dynflow/tasks/$TASK_ID"
+AUTH="$TASK_ID:$OTP"
+CURL="curl --silent --show-error --fail --max-time 10"
 
 # acquiring the lock before proceeding, to ensure only one instance of the script is running
 while ! mkfifo $BASE_DIR/retrieve_lock 1>/dev/null 2>&1; do
     # we failed to create retrieve_lock - assuming there is already another retrieve script running
     # waiting until it finished before we try to acquire it
     read -t 10 <>$BASE_DIR/retrieve_lock
 done
+
+release_lock() {
+    rm $BASE_DIR/retrieve_lock
+}
 # ensure the release the lock at exit
 trap "release_lock" EXIT
 
 pid=$(cat "$BASE_DIR/pid")
+[ -f "$BASE_DIR/position" ] || echo 1 > "$BASE_DIR/position"
+position=$(cat "$BASE_DIR/position")
+
+prepare_output() {
+    if [ -e $BASE_DIR/manual_mode ] || ([ -n "$pid" ] && pgrep -P "$pid" >/dev/null 2>&1); then
+        echo RUNNING
+    else
+        echo "DONE $(cat "$BASE_DIR/exit_code" 2>/dev/null)"
+    fi
+    [ -f "$BASE_DIR/output" ] || exit 0
+    tail --bytes "+${position}" "$BASE_DIR/output" > "$BASE_DIR/tmp"
+    cat "$BASE_DIR/tmp"
+}
+
+# prepare the callback payload
+payload() {
+    if [ -n "$1" ]; then
+        exit_code="$1"
+    else
+        exit_code=null
+    fi
+
+    if [ -e "$BASE_DIR/manual_mode" ]; then
+        manual_mode=true
+        output=$(prepare_output | base64 -w0)
+    else
+        manual_mode=false
+    fi
+
+    echo "{ \"step_id\": \"$STEP_ID\","\
+         "  \"manual_mode\": $manual_mode,"\
+         "  \"exit_code\": $exit_code,"\
```

Comment:
Updated: I've changed the order. Given the user of this script could actually just source the `env.sh` file and do whatever requests it would allow him, I don't think it makes much more sense to be cautions here, as it's still just a client-side validation. The main validation is happening on serve side.